File: src/modules/metrics/MetricsCollector.cpp
#include "MetricsCollector.h"
#include "../../core/Platform.h"
#include <sstream>
#include <iomanip>

MetricsCollector::MetricsCollector(Platform* platform)
    : platform_(platform) {
}

MetricsCollector::~MetricsCollector() {
}

void MetricsCollector::UpdateCounters() {
    if (platform_) {
        platform_->updatePdhMetrics();
    }
}

MetricsCollector::SystemMetrics MetricsCollector::CollectAll() {
    SystemMetrics metrics;

    metrics.performance = CollectPerformance();
    metrics.gpu = CollectGpu();
    metrics.processes = CollectProcesses();
    metrics.updates = CheckUpdates();

    return metrics;
}

MetricsCollector::PerformanceMetrics MetricsCollector::CollectPerformance() {
    PerformanceMetrics metrics;

    if (!platform_) {
        return metrics;
    }

    if (!cachingEnabled_) {
        platform_->invalidateMetricCaches();
    }

    metrics.cpuUsage = platform_->getCpuUsagePercent();
    metrics.ramUsage = platform_->getRamUsagePercent();
    metrics.diskQueue = platform_->getDiskQueueLength();
    metrics.netRetrans = platform_->getNetworkRetransRate();
    metrics.uptime = platform_->getSystemUptime();

    // Default to C: drive on Windows, root on Linux
#ifdef _WIN32
    metrics.freeDiskSpace = platform_->getFreeDiskSpaceGB("C:");
#else
    metrics.freeDiskSpace = platform_->getFreeDiskSpaceGB("/");
#endif

    return metrics;
}

MetricsCollector::GpuMetrics MetricsCollector::CollectGpu() {
    GpuMetrics metrics;

    if (!platform_) {
        return metrics;
    }

    if (!cachingEnabled_) {
        platform_->invalidateMetricCaches();
    }

    metrics.driverInfo = platform_->getGpuDriverInfo();
    metrics.usage = platform_->getGpuUsagePercent();

    return metrics;
}

MetricsCollector::ProcessMetrics MetricsCollector::CollectProcesses() {
    ProcessMetrics metrics;

    if (!platform_) {
        return metrics;
    }

    if (!cachingEnabled_) {
        platform_->invalidateMetricCaches();
    }

    metrics.highRamProcesses = platform_->getHighRamProcesses();

    return metrics;
}

MetricsCollector::UpdateStatus MetricsCollector::CheckUpdates() {
    UpdateStatus status;

    if (!platform_) {
        return status;
    }

    if (!cachingEnabled_) {
        platform_->invalidateMetricCaches();
    }

    status.state = platform_->getWindowsUpdateState();

    return status;
}

void MetricsCollector::InvalidateCache() {
    if (platform_) {
        platform_->invalidateMetricCaches();
    }
}

void MetricsCollector::SetCachingEnabled(bool enabled) {
    cachingEnabled_ = enabled;
}

std::string MetricsCollector::GetFormattedMetrics() {
    if (!platform_) {
        return "Error: No platform available";
    }

    SystemMetrics metrics = CollectAll();

    std::stringstream ss;
    ss << "=== System Metrics ===" << std::endl;
    ss << "CPU Usage:       " << metrics.performance.cpuUsage << "%" << std::endl;
    ss << "RAM Usage:       " << metrics.performance.ramUsage << "%" << std::endl;
    ss << "Disk Queue:      " << std::fixed << std::setprecision(2)
        << metrics.performance.diskQueue << std::endl;
    ss << "Net Retrans:     " << std::fixed << std::setprecision(2)
        << metrics.performance.netRetrans << " /s" << std::endl;
    ss << "Free Disk:       " << metrics.performance.freeDiskSpace << " GB" << std::endl;
    ss << "Uptime:          " << metrics.performance.uptime << std::endl;
    ss << std::endl;
    ss << "=== GPU Metrics ===" << std::endl;
    ss << "GPU Info:        " << metrics.gpu.driverInfo << std::endl;
    ss << "GPU Usage:       " << std::fixed << std::setprecision(1)
        << metrics.gpu.usage << "%" << std::endl;
    ss << std::endl;
    ss << "=== System Status ===" << std::endl;
    ss << "Update State:    " << metrics.updates.state << std::endl;
    ss << "High RAM Procs:  " << metrics.processes.highRamProcesses << std::endl;

    return ss.str();
}

File: rust/src/modules/metrics/metrics_collector.rs
use crate::core::platform::Platform;
use std::sync::{Arc, Mutex};

#[derive(Debug, Clone, Default)]
pub struct PerformanceMetrics {
    pub cpu_usage: i32,
    pub ram_usage: i32,
    pub disk_queue: f32,
    pub net_retrans: f32,
    pub uptime: String,
    pub free_disk_space: String,
}

#[derive(Debug, Clone, Default)]
pub struct GpuMetrics {
    pub driver_info: String,
    pub usage: f32,
}

#[derive(Debug, Clone, Default)]
pub struct ProcessMetrics {
    pub high_ram_processes: String,
}

#[derive(Debug, Clone, Default)]
pub struct UpdateStatus {
    pub state: String,
}

#[derive(Debug, Clone, Default)]
pub struct SystemMetrics {
    pub performance: PerformanceMetrics,
    pub gpu: GpuMetrics,
    pub processes: ProcessMetrics,
    pub updates: UpdateStatus,
}

pub struct MetricsCollector {
    platform: Arc<Mutex<Box<dyn Platform>>>,
    caching_enabled: Mutex<bool>,
}

impl MetricsCollector {
    pub fn new(platform: Arc<Mutex<Box<dyn Platform>>>) -> Self {
        Self {
            platform,
            caching_enabled: Mutex::new(true),
        }
    }

    pub fn update_counters(&self) {
        let mut p = self.platform.lock().unwrap();
        p.update_pdh_metrics();
    }

    pub fn collect_all(&self) -> SystemMetrics {
        SystemMetrics {
            performance: self.collect_performance(),
            gpu: self.collect_gpu(),
            processes: self.collect_processes(),
            updates: self.check_updates(),
        }
    }

    pub fn collect_performance(&self) -> PerformanceMetrics {
        let mut p = self.platform.lock().unwrap();

        if !*self.caching_enabled.lock().unwrap() {
            p.invalidate_metric_caches();
        }

        let free_disk_space = if cfg!(target_os = "windows") {
            p.get_free_disk_space_gb("C:")
        } else {
            p.get_free_disk_space_gb("/")
        };

        PerformanceMetrics {
            cpu_usage: p.get_cpu_usage_percent(),
            ram_usage: p.get_ram_usage_percent(),
            disk_queue: p.get_disk_queue_length(),
            net_retrans: p.get_network_retrans_rate(),
            uptime: p.get_system_uptime(),
            free_disk_space,
        }
    }

    pub fn collect_gpu(&self) -> GpuMetrics {
        let mut p = self.platform.lock().unwrap();

        if !*self.caching_enabled.lock().unwrap() {
            p.invalidate_metric_caches();
        }

        GpuMetrics {
            driver_info: p.get_gpu_driver_info(),
            usage: p.get_gpu_usage_percent(),
        }
    }

    pub fn collect_processes(&self) -> ProcessMetrics {
        let mut p = self.platform.lock().unwrap();

        if !*self.caching_enabled.lock().unwrap() {
            p.invalidate_metric_caches();
        }

        ProcessMetrics {
            high_ram_processes: p.get_high_ram_processes(),
        }
    }

    pub fn check_updates(&self) -> UpdateStatus {
        let mut p = self.platform.lock().unwrap();

        if !*self.caching_enabled.lock().unwrap() {
            p.invalidate_metric_caches();
        }

        UpdateStatus {
            state: p.get_windows_update_state(),
        }
    }

    pub fn invalidate_cache(&self) {
        let mut p = self.platform.lock().unwrap();
        p.invalidate_metric_caches();
    }

    pub fn set_caching_enabled(&self, enabled: bool) {
        let mut caching = self.caching_enabled.lock().unwrap();
        *caching = enabled;
    }

    pub fn get_formatted_metrics(&self) -> String {
        let metrics = self.collect_all();

        format!(
            "=== System Metrics ===\n\
            CPU Usage: {}%\n\
            RAM Usage: {}%\n\
            Disk Queue Length: {:.2}\n\
            Network Retransmission Rate: {:.2}\n\
            System Uptime: {}\n\
            Free Disk Space: {} GB\n\
            === GPU Metrics ===\n\
            GPU Driver Info: {}\n\
            GPU Usage: {:.2}%\n\
            === System Status ===\n\
            Update State: {}\n\
            High RAM Processes: {}\n",
            metrics.performance.cpu_usage,
            metrics.performance.ram_usage,
            metrics.performance.disk_queue,
            metrics.performance.net_retrans,
            metrics.performance.uptime,
            metrics.performance.free_disk_space,
            metrics.gpu.driver_info,
            metrics.gpu.usage,
            metrics.updates.state,
            metrics.processes.high_ram_processes
        )
    }
}